### DISCUSSION TITLE
Strictness of asset hierarchies

### DISCUSSION THREAD

**Discussion by tommyers-elastic** [elastic/assetbeat#29] [General]
Say we have the following AWS resources: 

EKS clusters, VPCs, Subnets, EC2 instances.

```
     ┌─────────────────────────────────────────────────┐
     │ VPC                                             │
     │                                                 │
     │ ┌─────────────────────────────┐                 │
     │ │ EKS Cluster                 │                 │
     │ │                             │                 │
     │ │ ┌────────────────────────┐  │                 │
     │ │ │  Subnet A              │  │                 │
     │ │ │                        │  │ ┌─────────────┐ │
     │ │ │ ┌────────┐ ┌────────┐  │  │ │  Subnet B   │ │
     │ │ │ │EC2     │ │EC2     │  │  │ │             │ │
     │ │ │ │Instance│ │Instance│  │  │ │ ┌────────┐  │ │
     │ │ │ │1       │ │2       │  │  │ │ │EC2     │  │ │
     │ │ │ └────────┘ └────────┘  │  │ │ │Instance│  │ │
     │ │ │                        │  │ │ │3       │  │ │
     │ │ └────────────────────────┘  │ │ └────────┘  │ │
     │ │                             │ │             │ │
     │ └─────────────────────────────┘ └─────────────┘ │
     │                                                 │
     └─────────────────────────────────────────────────┘
```

Let's say we can figure out which cluster and subnet an EC2 instance is in, and also which cluster it's in. **But let's say we can't figure out which cluster a subnet is running in at the asset level.** 

Whilst the EKS cluster is a parent of the instance, it's not the direct parent in the same way the subnet is. So how should we report these assets so we can figure out the hierarchy? With the explicit knowledge of how the picture looks we can figure it out, but ideally we don't rely on that kind of knowledge when building maps of asset relationships.

**Comment by tommyers-elastic** [comment]
@ruflin @jasonrhodes what are your thoughts on this?

**Reply by ruflin** [reply]
There are 3 properties: parents, children, references. There are clear parent / child relation chips like node / pods but in cases like network I think of it much more like a reference. I assume an EC2 instance could change from one subnet into another one. I see `subnet` in the above as a reference, not a parent. 

Same for VPC, I don't think every EC2 instance has to be in a VPC? Same for EC2 in an EKS cluster, it can be outside like you have shown. What is always inside a EKS cluster, is an k8s node. That brings up the question, can single asset have 2 roles? Or should we persist these assets just twice and indicate that these might be the same?

**Reply by jasonrhodes** [reply]
My sense is that "parent" and "child" describe relationships where one thing is "inside" of another thing. If an EC2 instance runs "inside" of a subnet, I'd consider it to be a child of that subnet. 

I wonder if the best thing to do might be to broaden the understanding of "parents" and "children" to mean "ancestors" and "descendants" and allow each asset document to list any combination of assets that fit those descriptions. For example:

```
# EC2 Instance 1
asset.parents: ["subnet A"]

# EC2 Instance 2
asset.parents: ["subnet A", "EKS Cluster"]

# EC2 Instance 3
asset.parents: ["subnet B", "VPC"]
```

Technically, all of these could be valid, and we just need to adjust our assumption that the assets that are described in `asset.parents` are always direct parents of this asset. I imagine that assumption will be wrong most of the time, anyway, because of the existence of interstitial assets we may not even be collecting data on at first, which later appear, so this is probably a smart way to go.

Consequences? These fields would no longer represent "an adjacency list" in graph terms, so it may eventually limit the usefulness of the information in these fields, but I'm not enough of a graph theory expert to know for sure. In the shorter term, it would just mean that as we manually traverse this graph, searching for an asset ID in the `asset.parents` field would indicate that the ID you are looking for is a descendant, not a direct child, of the other asset (vice versa with `asset.children`).

**Reply by jasonrhodes** [reply]
Another example, to help me think this through:

```
# Container 1 (runs in Pod A)
asset.parents: ["podA", "clusterXYZ"]

# Container 2 (runs in Pod B)
asset.parents: ["podB"]

# Container 3 (runs in Pod C)
asset.parents: ["clusterXYZ"]

# Pod A (runs in Node 10)
asset.parents: ["node10"]

# Pod B (runs in Node 11)
asset.parents: ["clusterXYZ"]

# Pod C (runs in Node 11)
asset.parents: ["clusterXYZ", "node11"]

# Node 10
asset.parents: ["clusterXYZ"]

# Node 11
asset.parents: ["clusterXYZ"]
```

In a generalized container view, with a single query I can show multiple related assets that exist "up the stack" (parents).

| Container ID | Parents |
| :--- | :--- |
| 1 | Pod A, Cluster XYZ |
| 2 | Pod B |
| 3 | Cluster XYZ |

If you want to show "all containers that run in cluster XYZ", you can query for:
```
[
  {
    "term": {
      "asset.parents": "clusterXYZ"
    }
  }
]
```
In the above example, this would return:
- Container 1
- Container 3
- Pod B
- Pod C
- Node 10
- Node 11

If you wanted to be more confident in the results, maybe you request it with a "distance" value of 3, which tells the API to do 2 more queries. First, remove the containers you already received, then query:

```json
{
  "should": [
    {
      "term": {
        "asset.parents": "Pod B"
      }
    },
    {
      "term": {
        "asset.parents": "Pod C"
      }
    },
    {
      "term": {
        "asset.parents": "Node 10"
      }
    },
    {
      "term": {
        "asset.parents": "Node 11"
      }
    }
  ],
  "minimum_should_match": 1
}
```
which would return:
- Container 2
- Pod A
- Pod C

With a distance of "3" you would do one more round:
```json
{
  "should": [
    {
      "term": {
        "asset.parents": "Pod A"
      }
    }
  ],
  "minimum_should_match": 1
}
```
(optimized in code so as to not search for the same parent twice)

That would again return Container 1. In the end, you'd be left with a list of all 3 containers running in the cluster. I need to think about this more but it might solve a number of fragility problems with how we were doing things when we assumed `asset.parents` and `asset.children` always represented direct relationships...

**Reply by jasonrhodes** [reply]
Last thought for now: this would still allow for `asset.references` to exist, but they would usually be "sibling" or "cousin" type relationships where the parent or child direction is unknown or irrelevant. Service G is related to Service Q, which uses `asset.references` to mark that relationship.

I also picture it being useful for something like an alert document, which may be able to use `asset.references` to list the asset IDs of the assets that a particular alert violation is referencing (however much of that information the alert framework knows at the time it detects the alert).

**Reply by ruflin** [reply]
I have a much stricture understanding of parent / children relation ship. Something is only a child if it can't exist without its parent. For example a pod must be in a node and a k8s node must be in a cluster (AFAIK). Everthing else is `references` / `related`. Also each asset should always only list its direct parent, not "grand parents". 

I don't expect the `asset` structure to magically allow to build a UI on top. The UI built must have domain knowledge about how different things relate built in on top fo the assets. Two UI's on could show different structures. Google has a good example of this in their Inventory view. A `k8s node` is also a `vm`. If you go to the `k8s node`, it shows pods, networks and things related which are specific to k8s. If you go to the `vm`, no pods are shown as in this UI, pods are not relevant. Ideally we connect these both UIs.

**Reply by jasonrhodes** [reply]
> Something is only a child if it can't exist without its parent.

I don't think this is different from what I've outlined here. The difference is whether the relationship must be direct or not, which the more I dig into real examples, the less I'm convinced we can be sure of this in all situations.

**Comment by jasonrhodes** [comment]
> Let's say we can figure out which cluster and subnet an EC2 instance is in, and also which cluster it's in.

Should one of those "cluster"s be "VPC"? I'm making sure I understand what you're describing before I weigh in.

**Comment by tommyers-elastic** [comment]
based on this discussion - i think the most simple way forward is to keep parent/child relationships strictly 'one-hop'. no ancestors or grandparents. we will put any of those relationships in 'references', and work with the UI team to build the topology maps correctly. 

on the question of should an asset be reported twice for different types e.g. k8s node and VM. **_i think we should do this._** again, the relationship will be 'reference', and based on this we can figure out how everything is linked.

@girodav @dmathieu @MichaelKatsoulis FYI

**Reply by jasonrhodes** [reply]
I'd like to revisit this question once we have a solid set of data in the indices and begin to look at the queries needed to drive the currently planned APIs. At the moment, it's not entirely clear the advantage of putting something in "references" if we know it's an ancestor or descendant, but perhaps relying on "references" won't be as complicated as I'm worried it could be.
